File: js/sms_service.js
/**
 * Twilio SMS Service for SENTINEL Web Dashboard
 * Handles sending SMS and OTPs via Twilio REST API
 */

const TWILIO_SID = 'ACa7209437c9427fdbce3c43808c22eb43';
const TWILIO_TOKEN = '629d0346ecab6f0ed7bcf2dd06602b8a';
const TWILIO_FROM = '+16168670252';

class SmsService {
    /**
     * Sends a general SMS
     * @param {string} to - Destination phone number
     * @param {string} body - Message content
     */
    static async sendSms(to, body) {
        console.log(`SENTINEL SMS: Sending to ${to}...`);
        
        // Ensure phone number starts with +
        if (!to.startsWith('+')) {
            to = '+91' + to; // Default to India if no prefix
        }

        const url = `https://api.twilio.com/2010-04-01/Accounts/${TWILIO_SID}/Messages.json`;
        const auth = btoa(`${TWILIO_SID}:${TWILIO_TOKEN}`);

        try {
            const response = await fetch(url, {
                method: 'POST',
                headers: {
                    'Authorization': `Basic ${auth}`,
                    'Content-Type': 'application/x-www-form-urlencoded'
                },
                body: new URLSearchParams({
                    'To': to,
                    'From': TWILIO_FROM,
                    'Body': body
                })
            });

            const data = await response.json();
            if (response.ok) {
                console.log("SENTINEL SMS: Success!", data.sid);
                return { success: true, sid: data.sid };
            } else {
                console.error("SENTINEL SMS: API Error", data.message);
                return { success: false, error: data.message };
            }
        } catch (e) {
            console.error("SENTINEL SMS: Network Error", e);
            return { success: false, error: e.message };
        }
    }

    /**
     * Sends a verification OTP
     * @param {string} to - Destination phone number
     */
    static async sendOtp(to) {
        const otp = Math.floor(100000 + Math.random() * 900000);
        const body = `[SENTINEL] Your emergency verification code is: ${otp}. Do not share this code.`;
        
        const result = await this.sendSms(to, body);
        if (result.success) {
            return { success: true, otp: otp };
        }
        return result;
    }
}

// Expose to window for sos_admin.js
window.SmsService = SmsService;

File: flutter_sos_app/lib/services/sms_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter/foundation.dart';

class SmsService {
  // TWILIO CREDENTIALS
  static const String _accountSid = 'ACa7209437c9427fdbce3c43808c22eb43';
  static const String _authToken = 'b0a8a2a1dce1b1b8da5b41b5fb0709f9';
  static const String _twilioNumber = '+16168670252';

  /// Sends an SMS via Twilio REST API
  /// [phone] should be in E.164 format (e.g., +919361488694)
  static Future<bool> sendSms(String phone, String message) async {
    // Ensure phone is in E.164 format
    String targetPhone = phone.startsWith('+') ? phone : '+$phone';

    final url = Uri.parse(
        'https://api.twilio.com/2010-04-01/Accounts/$_accountSid/Messages.json');

    try {
      final response = await http.post(
        url,
        headers: {
          'Authorization':
              'Basic ${base64Encode(utf8.encode('$_accountSid:$_authToken'))}',
        },
        body: {
          'From': _twilioNumber,
          'To': targetPhone,
          'Body': message,
        },
      );

      debugPrint(
          "Twilio HTTP Response: ${response.statusCode} - ${response.body}");
      return response.statusCode == 201;
    } catch (e) {
      debugPrint("Twilio API Connection Error: $e");
      return false;
    }
  }

  /// Sends a verification OTP
  static Future<bool> sendOtp(String phone, String code) async {
    final message = 'Your SENTINEL SOS verification code is: $code';
    return sendSms(phone, message);
  }

  /// Sends an emergency SOS alert
  static Future<bool> sendSosAlert({
    required String name,
    required String phone,
    required String description,
    required double latitude,
    required double longitude,
    String? emergencyContact,
  }) async {
    final message = '🆘 EMERGENCY SOS: $name ($phone) needs help!\n'
        'Location: https://www.google.com/maps?q=$latitude,$longitude\n'
        'Description: $description';
    
    // If no emergency contact provided, notify a central number or the user themselves (for testing)
    // In a real app, this would be an admin or a local authority number.
    final target = emergencyContact ?? phone; 
    return sendSms(target, message);
  }
}
